File: src/main/java/org/c3lang/intellij/intention/ImportModuleInspection.java
package org.c3lang.intellij.intention;

import com.intellij.codeInspection.LocalInspectionTool;
import com.intellij.codeInspection.LocalInspectionToolSession;
import com.intellij.codeInspection.ProblemHighlightType;
import com.intellij.codeInspection.ProblemsHolder;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.psi.PsiElementVisitor;
import com.intellij.psi.util.PsiTreeUtil;
import org.c3lang.intellij.completion.ImportCompletionContributor;
import org.c3lang.intellij.index.NameIndexService;
import org.c3lang.intellij.psi.C3BaseType;
import org.c3lang.intellij.psi.C3CallExpr;
import org.c3lang.intellij.psi.C3Expr;
import org.c3lang.intellij.psi.C3File;
import org.c3lang.intellij.psi.C3FullyQualifiedNamePsiElement;
import org.c3lang.intellij.psi.C3ModuleDefinition;
import org.c3lang.intellij.psi.C3Path;
import org.c3lang.intellij.psi.C3PathIdentExpr;
import org.c3lang.intellij.psi.C3Visitor;
import org.c3lang.intellij.psi.ModuleName;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

public class ImportModuleInspection extends LocalInspectionTool
{
    private static final Logger LOG = Logger.getInstance(ImportCompletionContributor.class);

    @Override
    public @NotNull String getDisplayName()
    {
        return "Import module";
    }

    @Override
    public @NotNull String getGroupDisplayName()
    {
        return "Import";
    }

    @Override
    public @NotNull PsiElementVisitor buildVisitor(
            @NotNull ProblemsHolder holder,
            boolean isOnTheFly,
            @NotNull LocalInspectionToolSession session)
    {
        if (!(session.getFile() instanceof C3File))
        {
            return PsiElementVisitor.EMPTY_VISITOR;
        }

        return new C3Visitor()
        {
            @Override
            public void visitCallExpr(@NotNull C3CallExpr psi)
            {
                C3Expr expr = psi.getExpr();
                if (expr instanceof C3PathIdentExpr pathIdentExpr && pathIdentExpr.getPathIdent().getPath() != null)
                {
                    registerPathProblem(pathIdentExpr.getPathIdent().getPath());
                }
            }

            @Override
            public void visitBaseType(@NotNull C3BaseType o)
            {
                if (o.getPath() != null)
                {
                    registerPathProblem(o.getPath());
                }
            }

            @Override
            public void visitPath(@NotNull C3Path o)
            {
                registerPathProblem(o);
            }

            private void registerPathProblem(@NotNull C3Path psi)
            {
                Collection<C3FullyQualifiedNamePsiElement> callables =
                    NameIndexService.INSTANCE.findByNameEndsWith(psi.getParent().getText(), psi.getProject());
                List<C3FullyQualifiedNamePsiElement> callableList = new ArrayList<>(callables);
                C3FullyQualifiedNamePsiElement element = callableList.size() == 1 ? callableList.get(0) : null;

                if (element == null)
                {
                    String details = String.join(", ", callableList.stream()
                        .map(callable -> callable.getFqName().getFullName())
                        .toList());
                    LOG.debug("Ignoring QuickFix, callables found: " + details);
                    return;
                }

                C3ModuleDefinition importProvider = java.util.Objects.requireNonNull(
                    PsiTreeUtil.getParentOfType(psi, C3ModuleDefinition.class)
                );

                ModuleName importIntention = element.getFqName().getModule();
                if (importIntention == null) return;

                if (importIntention.getValue().equals("std::core") || importIntention.getValue().startsWith("std::core::"))
                {
                    return;
                }

                if (importProvider.containsImportOrSameModule(element)) return;

                ModuleName applied = psi.getUserData(AddImportQuickFix.KEY);
                psi.putUserData(AddImportQuickFix.KEY, null);
                if (applied != null) return;

                holder.registerProblem(
                    psi,
                    "Import " + importIntention.getValue(),
                    ProblemHighlightType.WEAK_WARNING,
                    new AddImportQuickFix(psi, importIntention)
                );
            }
        };
    }
}

File: src/main/java/org/c3lang/intellij/intention/ImportModuleInspection.kt
package org.c3lang.intellij.intention

import com.intellij.codeInspection.LocalInspectionTool
import com.intellij.codeInspection.LocalInspectionToolSession
import com.intellij.codeInspection.ProblemHighlightType
import com.intellij.codeInspection.ProblemsHolder
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.util.removeUserData
import com.intellij.psi.PsiElementVisitor
import com.intellij.psi.util.parentOfType
import org.c3lang.intellij.completion.ImportCompletionContributor
import org.c3lang.intellij.index.NameIndexService
import org.c3lang.intellij.psi.*


class ImportModuleInspection : LocalInspectionTool() {
    private val log = Logger.getInstance(
        ImportCompletionContributor::class.java
    )

    override fun getDisplayName(): String = "Import module"

    override fun getGroupDisplayName(): String = "Import"

    override fun buildVisitor(
        holder: ProblemsHolder,
        isOnTheFly: Boolean,
        session: LocalInspectionToolSession
    ): PsiElementVisitor {
        if (session.file !is C3File) {
            return PsiElementVisitor.EMPTY_VISITOR
        }
        return object : C3Visitor() {

            override fun visitCallExpr(psi: C3CallExpr) {
                when (val expr = psi.expr) {
                    is C3PathIdentExpr -> expr.pathIdent.path?.let { registerPathProblem(it) }
                }
            }

            override fun visitBaseType(o: C3BaseType) {
                o.path?.let { registerPathProblem(it) }
            }

            override fun visitPath(o: C3Path) {
                registerPathProblem(o)
            }

            private fun registerPathProblem(psi: C3Path) {
                val callables = NameIndexService.findByNameEndsWith(psi.parent.text, psi.project)
                val element = callables.singleOrNull()

                if (element == null) {
                    val details = callables.joinToString(", ") { it.fqName.fullName }
                    log.debug("Ignoring QuickFix, callables found: {}", details)
                    return
                }
                val importProvider = checkNotNull(psi.parentOfType<C3ModuleDefinition>())

                val importIntention = element.fqName.module ?: return

                if (importProvider.imports.contains(importIntention)) {
                    return
                }

                val applied = psi.removeUserData(AddImportQuickFix.KEY)
                if (applied == null) {
                    holder.registerProblem(
                        psi,
                        "Import ${importIntention.value}",
                        ProblemHighlightType.WEAK_WARNING,
                        AddImportQuickFix(
                            target = psi,
                            importIntention = importIntention,
                        )
                    )
                }
            }
        }
    }

}